### PR TITLE
Added a new 'velocity_scaling_factor' parameter feature to joint_limits.yaml

### PR DESCRIPTION
**Pull request by davetcoleman**
This evenly reduces max joint velocity for all joints. It is particularly useful if you are using the default max_velocities from the URDF, but want to slow down your robot say by 10% (set it to 0.1)

As we discussed @mmurooka

This PR is documented via the Setup Assistant https://github.com/ros-planning/moveit_setup_assistant/pull/92

### CODE REVIEW

**Comment by adolfo-rt** [comment]
Currently the `joint_limits.yaml` already allows to override velocity limits specified in the URDF. Isn't this new parameter redundant?.

Since you mention the Setup Assistant, it _could_ be convenient if there was a mechanism (in the assistant) to simplify the specification of the velocity limits by allowing it to scale down the URDF values, but I wouldn't use the `joint_limits.yaml` file for that.

**Comment by davetcoleman** [comment]
This new parameter is a _global_ scale factor, so you just edit one parameter and the whole robot essentially changes speed. This is something I commonly need to tweak when debugging. It saves me from having to use a calculator to edit every joint value by hand.

I see your point about duplicate functionality, but as a command-line oriented person I like to be able to quickly edit a text file. Also, the mechanism in the assistant would be nice, but would also take much more work to implement and it would require a new screen. 

To clarify, my joint_limits.yaml now looks like:

```
joint_limits:
  velocity_scaling_factor: 0.1
  CHEST_JOINT0:
    has_velocity_limits: true
    max_velocity: 4.40217
    has_acceleration_limits: false
    max_acceleration: 0
...
```

**Comment by adolfo-rt** [comment]
Well, considering that joint limits are not dynamic in the sense that you need to relaunch your MoveIt-using application for the change to be effective, you could just as well have two joint limit specs and switch between them when debugging. A symlink that points to the relevant file would allow you to not change your launch files, i.e., `joint_limits.yaml -> joint_limits_slow.yaml`.

Two additional things I don't like much about the patch are:
- Why limit this feature to velocity limits only?. One might expect scaling to be available for other variables, like acceleration.
- All code that parses this specification must now be made aware of the new parameter to have consistent behavior across the board. At this moment we don't know how many people leverage the joint limits specification outside `moveit_ros`, but at least I do.

I think your usecase is legitimate, I'm just not convinced of the proposed solution.

**Comment by davetcoleman** [comment]
I thought about adding the acceleration scaling factor, but decided not to yet since I didn't need it. I'd be happy to add it.

I wasn't aware others were using the spec, but I'd point out this doesn't really break backwards compatibility. 

I agree there are plenty of better solutions, but this was a quick fix that is better than hard-coding the value in for my own application (which I debated). I've heard so many people ask how slow down the robot with MoveIt! that I thought it would be useful. If we had infinite developers and time we could do all sorts of better things:
- Setup Assistant GUI interface
- Rviz Plugin GUI interface
- Dynamic Reconfigure control

Basically, MoveIt! needs a speed knob and this is the closest its ever had. Given we don't have anything better, I'm not 100% convinced its a bad idea.

**Comment by skohlbr** [comment]
One of the reasons we implemented our own move_group capability was actually to be able to specify a velocity scaling factor on a per motion request basis. Ideally this would be part of the MotionPlanRequest message IMHO. It´s actually interesting that with all the constraints that can be specified, velocity doesn´t seem to be one them (?).

**Comment by davetcoleman** [comment]
+1 to velocity scaling factor in MotionPlanRequest, that would be a great addition. 

How about send a PR to add that feature? We'd also need to make a few small changes in MoveIt! besides the message, I'd be happy to help.

**Comment by shaun-edwards** [comment]
+1....We'd like to see this as well.

**Comment by sachinchitta** [comment]
The velocity limits are used in only one place inside move_group itself - the iterative filter that parameterizes the trajectory - they are currently not used for the planning (although they could be). I would be fine adding a velocity scaling factor in there so this could happen on a per request basis. Note that this will change the MotionPlanRequest message definition - the change is not backwards compatible. Also, this will require more than just a few changes in MoveIt! - currently the velocity and acceleration bounds come into the RobotModel class and are used from there. The scaling factor will have to apply when they are read. I would certainly not have the scaling factor be a parameter in joint_limits.yaml.

A default scaling factor of 0.0 is a problem for the message structure though, this will break all existing MoveIt! calls - some way of getting around that would be nice - could be a flag that says use the scaling factor or not. Also, we might need factors for both velocity and acceleration.  I will entertain a proposed PR but prefer it also includes regression tests that make sure we are not breaking backwards compatibility.

**Comment by davetcoleman** [comment]
>  I would certainly not have the scaling factor be a parameter in joint_limits.yaml

Ok, I understand. Thanks @adolfo-rt for the lively discussion

>  the change is not backwards compatible. 

I still think we need a better way to allow improvements to MoveIt! without always breaking people's code who don't want it.

@skohlbr I'd really like a replacement for this velocity scaling favor, MoveIt! needs it badly. What do you think?

**Comment by skohlbr** [comment]
@davetcoleman Definitely agree. So putting up for discussion what we did so far (which is a simple approach, but it worked well for us). Essentially, after planning we would call the retimeTrajectory method below and then send the retimed trajectory out for execution.

```
void move_group::CustomMoveGroupPlugin::retimeTrajectory(const trajectory_msgs::JointTrajectory& from, trajectory_msgs::JointTrajectory& to, double factor)
{
  to = from;

  size_t num_traj_points = to.points.size();
  size_t num_joints = to.joint_names.size();

  double inv_factor = (1.0/factor);

  for (size_t i ; i < num_traj_points; ++i){
    to.points[i].time_from_start *= inv_factor;

    for (size_t j; j < num_joints; ++j){
      to.points[i].velocities[j] *= factor;
    }
  }
}
```

We haven´t added the factor to the MotionPlanRequest yet (but set it externally), so not sure about all the implications of that. It appears to me that using default timing for (factor <= 0.0) and (factor >= 1.0) should make things behave as before for people using MotionPlanRequest with the new velocity factor (but not setting it), or am I missing something?

**Comment by adolfo-rt** [comment]
One last, semi-off-topic comment to keep an eye on the longer term vision. The _speed knob_ is not just something convenient for debugging, but it's a requirement for compliance with recent collaborative and service robot standards, like **ISO 13482:2014** _Robots and robotic devices — Safety requirements for personal care robots_.

In this context, speed should be continuously monitored and controlled as a function of the task type and user proximity. At some point I wanted to implement a speed throttling feature in ros_controller's `joint_trajectory_controller`, but never got around it. It's more work in the end, both on the MoveIt! and the controller sides, but I see a lot of value in this approach. It would even allow to pause and resume motions, or go backwards.

**Comment by skohlbr** [comment]
Note very similar and relevant discussion in #368

**Comment by skohlbr** [comment]
I implemented a scaling factor that can be specified in the MotionPlanRequest in these commits:
https://github.com/team-vigir/moveit_msgs/commit/08aa1f1735b1f31d492b676dab2506d487c5573f
https://github.com/team-vigir/moveit_ros/commit/df6e8303134cf95a35edbde7f3c69b4ee1c7f75a
https://github.com/team-vigir/moveit_core/commit/5552468e0141d5258103b393d271fed2d42d7119

The "motion_execution_time_factor" has a permissible range of (0,1]. If not within that range, it gets set to 1.0 in iterative time parametrization automatically, so behavior should be compatible with earlier moveit versions (to some extend, the MotionPlanRequest has to be updated of course).

Associated Pull Request(s): https://github.com/ros-planning/moveit_ros/pull/547

**Comment by davetcoleman** [comment]
Thanks!
